pkg/proctree/fork_processing.go

```diff
@@ -0,0 +1,176 @@
+package proctree
+
+import (
+	"github.com/aquasecurity/tracee/pkg/utils/types"
+	"github.com/aquasecurity/tracee/signatures/helpers"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+type taskIds struct {
+	Pid  int
+	Tid  int
+	Ppid int
+}
+
+// ProcessForkEvent add new process to process tree if new process created,
+// or update process threads if new thread created.
+// Because the fork at start is only a copy of the father,
+// the important information regarding of the process information and binary will be collected
+// upon execve.
+func (tree *ProcessTree) ProcessForkEvent(event *trace.Event) error {
+	err := tree.processGeneralEvent(event)
+	if err != nil {
+		return err
+	}
+
+	newHostIds, err := parseForkHostIds(event)
+	if err != nil {
+		return err
+	}
+	newNsIds, err := parseForNsIds(event)
+	if err != nil {
+		return err
+	}
+
+	if newHostIds.Pid == newHostIds.Tid {
+		err = tree.addForkProcess(event, newHostIds, newNsIds)
+		if err != nil {
+			return err
+		}
+	}
+	return tree.addForkThread(event, newHostIds, newNsIds)
+}
+
+// addForkProcess create a new process using fork event and the IDs given to the new process
+// Notice that the new process information is a duplicate of the father, until an exec will occur.
+func (tree *ProcessTree) addForkProcess(event *trace.Event, newInHostIds taskIds, newInNsIds taskIds) error {
+	newProcess, err := tree.getProcess(newInHostIds.Pid)
+	// If it is a new process or if for some reason the existing process is a result of lost exit
+	// event
+	if err != nil ||
+		newProcess.getForkTime() != 0 {
+		tree.immediateRemoveProcessFromTree(newInHostIds.Pid)
+		newProcess, err = tree.newForkedProcessNode(newInHostIds, newInNsIds)
+		if err != nil {
+			return err
+		}
+	}
+	newProcess.mutex.Lock()
```

Comment:
This is a no for me. The synchronization mechanism has to be controled by the interface/class/struct that contains it. Or else things will get out of control in no time (and the encapsulation characteristic of object orientation vanishes).